[WIP] Refactor testsuite setup

jougs opened the pull request:
This PR refactors the test runner scripts of NEST and PyNEST and makes them independent of configuration-time variables previously replaced into the scripts.

Changes in detail:
* `do_tests.sh` now gets all paths and executables on the commandline (fixes #345).
* `nest_vars.sh` does not export paths anymore, that could have confused NEST(fixes #972).
* Tests are now run in parallel with as many processors as are available (kudos to @terhorstd).
* Test results fro SLI tests are not written as JUnit/XML anymore.
* The info output from the `RandomManager` initialization is gone (fixes #700 and #415).
* The PyNEST testsuite was
   * moved outside of the `nest` module directory
   * converted to using `pytest` instead of `nosetests`
   * modified so that PyNEST tests do not include MPI anymore (fixes #1130).
* Increased legibility by adding an informative header and removing redundant paths from the testsuite printouts.

heplesser commented:
@jougs In the Python-free test configurations a `*.py` regression test fails.

jougs commented:
@stinebuu: can you please check if [what I did to your script from #1120](https://github.com/nest/nest-simulator/pull/1071/files#diff-e8570780ae94f31e498d275238e8083e) is good (or reasonable, at least :wink:)?

@heplesser, @gtrensch: I think I've sorted out all remaining issues with failing tests, formatting and such. As a bonus for the long wait, @terhorstd added full parallelization of the testsuite runs to this PR. Feel free to start reviewing at any time now.

stinebuu commented on pynest/nest/tests/test_sp/test_mpitests.py:
```diff
@@ -19,45 +19,41 @@
 
-HAVE_MPI = nest.ll_api.sli_func("statusdict/have_mpi ::")
+cmd = ["nest", "-c", "statusdict/have_mpi :: =only"]
+HAVE_MPI = subprocess.check_output(cmd) == "true"
```
@jougs I think `subprocess.check_output()` returns data as encoded bytes for Python 3 and as regular strings for Python 2. The test will only run if I use Python 2 and when I use Python 3 I get the message `skipped 'NEST was compiled without MPI'`, even though I have compiled with MPI. The same is true for `test_connect_all_patterns`. The two tests are also skipped during `make installcheck`.

heplesser commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+echo
+echo "  NEST testsuite"
+echo "  Date: $(date -u)"
+echo "  Sysinfo: $(uname -s -r -m -o)"
```
`uname`  on OSX does not support `-o`, I'd suggest to remove it.

terhorstd replied:
Would `-a` work on OSX?
```suggestion
echo "  Sysinfo: $(uname -a)"
```
I'm not sure if there was a reason against that…

heplesser commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+INFO_OS="$(uname -s)"
+INFO_USER="$(whoami)"
+INFO_VER="$(uname -r)"
+NPROCS="$(cat /proc/cpuinfo | grep processor | wc -l)"
```
`/proc/cpuinfo` does not exist on OSX. Under OSX, `sysctl -n hw.physicalcpu` returns the number of processors. 
Will cpuinfo work on more exotic systems?

terhorstd replied:
I think the standard coreutils command for this is
```suggestion
NPROCS="$(nproc)"
```
Are coreutils available in OSX?

heplesser commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+
+print_paths () {
+    indent="`printf '%23s'`"
+    echo "$1" | sed "s/:/\n$indent/g" | sed '/^\s*$/d'
```
`sed` works differently on OSX since it is inherited from BSD, see eg https://unix.stackexchange.com/questions/13711/differences-between-sed-on-mac-osx-and-other-standard-sed

One solution is to require the user to install GNU sed as `gsed` (e.g. via `brew install gnu-sed`) and defining the `sed` program to use as a variable depending on OS.

heplesser commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+    fi
+}
+
+phase_one
```
Should we add to the name what the tests are doing? E.g. `phase_one_selftests`?

jougs replied:
I think that we should actually have some kind of associative array mapping phase names to phase functions and then automatic command line arguments along the lines of `--list-phases`and `--skip-phases`. Ideally the phase functions would be in the corresponding directories, accompanied by a metadata file. I would, however, find this to be out of the scope of this PR.

@terhorstd: what is your take on this?

heplesser replied:
@jougs I like your idea! It would be useful to be able to selectively run only certain groups of tests.

jougs replied:
@heplesser: thanks for liking the idea. Should this also go into this PR or should it be its own?

heplesser replied:
@jougs There is probably still enough to do in this PR, so I wouldn't want to enlarge. Create a separate one for the named and flexible phases.

heplesser commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+    #    for test_dir in "unittests/" "topology/unittests/" ; do
+    #        for test_name in $(ls "${TEST_BASEDIR}/${test_dir}" | grep ".*\.${test_ext}\$") ; do
+    echo "TEST_BASEDIR $TEST_BASEDIR"
+    echo "PWD $(pwd)"
```
Remove these two output lines and the commented-out code above.

heplesser commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+echo "  Total number of tests: $(( $(cat "${TEST_TOTAL}" | wc -l) + $CPP_TEST_TOTAL + $PYNEST_TEST_TOTAL))"
+echo "     Passed: $(( $(cat "${TEST_PASSED}" | wc -l) + $CPP_TEST_PASSED + $PYNEST_TEST_PASSED ))"
+echo "     Skipped: $(( $(cat "${TEST_SKIPPED}" | wc -l) + $CPP_TEST_SKIPPED + $PYNEST_TEST_SKIPPED )) ${PYNEST_TEST_SKIPPED_TEXT:-}"
+echo "     Failed: $(( $(cat "${TEST_FAILED}" | wc -l) + $CPP_TEST_FAILED + $PYNEST_TEST_FAILED )) ${PYNEST_TEST_FAILED_TEXT:-}"
```
- Could one report total time used here?
- Could one provide info on which tests failed?

heplesser commented on testsuite/run_test:
```diff
@@ -0,0 +1,175 @@
+
+msg_error=
+
+script_name="$( echo "${param_script}" | sed 's/^.*\/\([^/]\+\)$/\1/' )"
```
Does not work on OSX, must use `gsed` or modify expressions.

heplesser commented on testsuite/run_test:
```diff
@@ -0,0 +1,175 @@
+#
+time_cmd()
+{
+    start=`date +%s%N`
```
This will give the number of seconds with `N` appended on OSX and then arithmetic below works no more. `date +%s` gives just the seconds since the epoch, so by dropping the division below, one gets a coarse measure. Not sure how to get ns-resolution time on OSX.

jougs commented on nestkernel/rng_manager.cpp:
```diff
@@ -235,8 +235,6 @@ nest::RNGManager::create_rngs_()
   }
 
-  LOG( M_INFO, "Network::create_rngs_", "Creating default RNGs" );
-
```
See below.

jougs commented on nestkernel/rng_manager.cpp:
```diff
@@ -280,8 +278,6 @@ nest::RNGManager::create_rngs_()
 nest::RNGManager::create_grng_()
 {
-  // create new grng
-  LOG( M_INFO, "Network::create_grng_", "Creating new default global RNG" );
```
The output in this line and the one mentioned above make `regressiontests/issue-779-1016.py` fail on my machine (the test captures output and compares it to some reference).

As the error only started since the changes in this PR I thought that problem might also affect other users and removed them (they were not that helpful, were they?).

heplesser replied:
@jougs This is a bit strange. The test ([see here](https://github.com/nest/nest-simulator/blob/2c9d73f2306d391902719c8bc88c851811869b7f/testsuite/regressiontests/issue-779-1016.py#L49)) should print the argv from statusdict after the kernel has completely started, so the CreateRNG messages should come first. Why do they come later on your machine? Interestingly, they cannot even be suppressed using `--verbosity=FATAL`, since they apparently are executed before verbosity has been handled.

But maybe these lines here should go, after all, `Create()` doesn't print a message "new neurons created" either.

jougs replied:
Argh, you are right!

I was pointing you to the wrong test and even got the context all wrong. I meant [test_connect_all_patterns.py](https://github.com/nest/nest-simulator/pull/1071/files#diff-5e7934b0517f89b433502dd81359c5f6R27) and [test_mpitests.py](https://github.com/nest/nest-simulator/pull/1071/files#diff-e8570780ae94f31e498d275238e8083eR27). I the linked lines I need to get the output of NEST used in pure command-mode. And as you correctly observed, these messages cannot be silenced at all, because they are printed from some constructor long before the verbosity setting is actually taken into account.

See also #415 and #700 (which I've now added to the description of this one).

jougs commented on pynest/do_tests.py:
```diff
@@ -66,7 +69,10 @@ def _setup(arg):
 
 total = result.testsRun + result_topo.testsRun
-failed = len(result.failures) + len(result.errors) + len(result_topo.failures) + len(result_topo.errors)
+failed = len(result.failures) + \
```
Resolved by in c0935ab5c

jougs commented:
@terhorstd: Can you please look into commented-out-code and portability-issues of the code you contributed? Also the one-report-per-test seems to come from your parallelization work.

Here are some additional hints:
* [SO#45181115](https://stackoverflow.com/questions/45181115) has a portable solution for getting the number of processors
* `do_tests.sh` contained a `portable_inplace_sed` function before your commit  f99d345d019dc7f815b1da3b378609dffedf1a71 maybe that helps in solving the portability problems.

jougs commented:
@stinebuu: Thanks for catching the Python 3 issue and that the test was not actually run and thus could not fail. I've fixed this in b1bcbf7. Can you please re-evaluate?

jougs commented:
I'm afraid that I have to be more invasive to complete this. I've added the `[WIP]` label to the title of the PR in order to make people aware that this is still worked on. @gtrensch, @stinebuu, @heplesser: I'll notify you once I think the code is ready for re-review. Sorry for the noise.

jougs commented:
Getting there! Please note that although Travis is convinced that everything is good now, it isn't. The Pytest MPI tests are still failing and I have messed up error detection in the runner script. I hope to get this fixed tomorrow. Stay tuned!

jougs commented:
@stinebuu, @heplesser: As this got way bigger than expected, I now think that it might be a much better idea to merge this into `nest-3` instead of master. Do you have any objections against that? If not, I will close this PR and create a new one once I'm done with the rebasing. Thanks!

heplesser commented:
@jougs I am wary of changing the test setup at the same time as we make a massive change to NEST. How about first merging `nest-3` into master and then add this to master 2-3 weeks later, but definitely before releasing NEST3?

BTW, there is one feature I have been longing for for a long time: A concise overview over the tests (SLI or Py) that failed at the end, instead of the current "1 out of 900 test failed. Enjoy scrolling back to find which one!" situation. Any hope for such luxury?

jougs commented:
Alright. Back in business on this one :-)

As I have decided to rebase the whole thing onto nest-3 to make everything easier down the line, I will now close this PR and open a new one against that branch. Stay tuned!

jougs commented:
This PR was originally meant to refactor the test runner scripts of NEST and PyNEST to make them independent of configuration-time variables previously replaced into the scripts. Unfortunately, it suffered (and eventually died) from a severe case of _feature creep_. Mostly for the record, the following summary re-iterates the detailed list of changes in this PR, with annotations that describe the current state of affairs:

* `do_tests.sh` now gets all paths and executables on the commandline (fixes #345).
   * **fixed by replacement #1686**
* `nest_vars.sh` does not export paths anymore, that could have confused NEST(fixes #972).
   * **fixed by replacements #1561 and  #1578**
* Tests are now run in parallel with as many processors as are available (kudos to @terhorstd).
* Test results fro SLI tests are not written as JUnit/XML anymore.
   * **JUnit/XML is used to gather information from tests since #1587, rendering this change inappropriate** 
* The info output from the `RandomManager` initialization is gone (fixes #700 and #415).
   * **fixed by replacement #1577**
* The PyNEST testsuite was
   * moved outside of the `nest` module directory
   * converted to using `pytest` instead of `nosetests`
   * modified so that PyNEST tests do not include MPI anymore (fixes #1130).
* Increased legibility by adding an informative header and removing redundant paths from the testsuite printouts.
   * **fixed by replacement #1686**

jougs commented:
for the record: The only un-merged thing from this PR is the parallelization of test runs with as many processors as are available.

terhorstd commented on testsuite/run_test:
```diff
@@ -0,0 +1,175 @@
+
+# from http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -eu
+#set -o pipefail
```
The "from" line is a proper comment that describes the following line. I strongly vote for leaving that in.
Actually the line should be a single enabled line:
```suggestion
set -euo pipefail
```
Some constructs may however be sensitive to `pipefail`, so this needs an explicit look at the code to enable without surprises.

terhorstd commented on testsuite/do_tests.sh:
```diff
@@ -0,0 +1,684 @@
+
+# from http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -eu
+#set -o pipefail
```
same as above.
